### PR TITLE
SINGA 475 - add logical operator: and, or, xor, not & negative, reciprocal

### PR DESCRIPTION
**Pull request by pinpom**

### CODE REVIEW

File: python/singa/autograd.py [hunk 1]
```diff
@@ -2522,3 +2523,144 @@ def backward(self, dy):
 def max(a,b):
     return Max()(a,b)[0]
 
+
+class And(Operation):
+    def __init__(self):
+        super(And, self).__init__()
+
+    def forward(self, a, b):
+        m = singa.__div__(a,b)
```
Comment (joddiy) [inline python/singa/autograd.py]: This part cannot pass the cpu test case, plz use this.
```
        m = singa.__mul__(a, b)
        cur = singa.PowFloat(singa.Sign(m), 2)
```

File: python/singa/autograd.py [hunk 2]
```diff
@@ -2522,3 +2523,144 @@ def backward(self, dy):
...
+    return And()(a,b)[0]
+
+
+class Or(Operation):
+    def __init__(self):
+        super(Or, self).__init__()
+
+    def forward(self, a, b):
+        #find equal element-wise
+        m = singa.__sub__(a,b)
```
Comment (joddiy) [inline python/singa/autograd.py]: not clear, plz use this:
```
        m = singa.__add__(singa.PowFloat(singa.Sign(a), 2.0), singa.PowFloat(singa.Sign(b), 2.0))
        cur = singa.Sign(m)
```

File: python/singa/autograd.py [hunk 3]
```diff
@@ -2522,3 +2523,144 @@ def backward(self, dy):
...
+    return Not()(x)[0]
+
+
+class Xor(Operation):
+    def __init__(self):
+        super(Xor, self).__init__()
+
+    def forward(self, a, b):
+        #find element with value =0
+        m0 = singa.__mul__(a,b)
```
Comment (joddiy) [inline python/singa/autograd.py]: not clear, plz use this:
```
        m = singa.__sub__(singa.PowFloat(singa.Sign(a), 2.0), singa.PowFloat(singa.Sign(b), 2.0))
        cur = singa.PowFloat(singa.Sign(m), 2.0)
```

File: python/singa/autograd.py (hunk 2 above)
Comment (pinpom) [reply python/singa/autograd.py]: i think the equation is not correct if a = -b (element-wise)

Comment (joddiy) [reply python/singa/autograd.py]: if a = -b, the Signs of them are, 1, and -1, and then, you take a square on it, the result will be 1 and 1, and you add them together, is 2, take the Sign again, you get a 1 finally.

Comment (pinpom) [reply python/singa/autograd.py]: ok got it. just changed the code already.

**Pull request merged**